### PR TITLE
Add Optimized Shell Sort

### PR DESCRIPTION
**Pull request by ZicsX**
### Describe your change:
Added a new shell sort implementation which is slightly faster than its pure implementation.

This shell sort is implemented using a gap, which shrinks by a certain factor each iteration. In this implementation, the gap is initially set to the length of the collection. The gap is then reduced by a certain factor (1.3) each iteration. 

For each iteration, the algorithm compares elements that are a certain number of positions apart (determined by the gap). If the element at the higher position is greater than the element at the lower position, the two elements are swapped. The process is repeated until the gap is equal to 1.

The reason this is more efficient is that it reduces the number of comparisons that need to be made. 
By using a smaller gap, the list is sorted more quickly.

image"""

* [x] Add an algorithm?
* [ ] Documentation change?
* [ ] Fix a bug or typo in an existing algorithm?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by L3str4nge — APPROVED** [review]

**Comment by ZicsX** [comment]
algorithms-keeper review-all

**Comment by ZicsX** [comment]
@algorithms-keeper review-all

**Comment by ZicsX** [comment]
@algorithms-keeper review

**Review by poyea — APPROVED** [review]
Thank you for your pull request!🤩

**Comment by poyea** [comment]
```
2022-07-03T16:34:40.6973928Z ==================================== ERRORS ====================================
2022-07-03T16:34:40.6974259Z _______________ ERROR collecting quantum/ripple_adder_classic.py _______________
2022-07-03T16:34:40.6974616Z quantum/ripple_adder_classic.py:6: in <module>
2022-07-03T16:34:40.6975140Z     from qiskit.providers import BaseBackend
2022-07-03T16:34:40.6975800Z E   ImportError: cannot import name 'BaseBackend' from 'qiskit.providers' (/opt/hostedtoolcache/Python/3.10.5/x64/lib/python3.10/site-packages/qiskit/providers/__init__.py)
2022-07-03T16:34:40.6976280Z =============================== warnings summary ===============================
```
Seems unrelated.

File: DIRECTORY.md [hunk 1]
```diff
@@ -973,6 +973,7 @@
   * [Recursive Quick Sort](sorts/recursive_quick_sort.py)
   * [Selection Sort](sorts/selection_sort.py)
   * [Shell Sort](sorts/shell_sort.py)
+  * [Shrink Shell Sort](sorts/shrink_shell.py)
```
Comment (poyea) [inline DIRECTORY.md:976]: Let's run the directory generator once?

Comment (ZicsX) [reply DIRECTORY.md:976]: Yes, Ok

**Comment by ZicsX** [comment]
Hi @poyea, please review and help me merge this change..

**Comment by poyea** [comment]
@ZicsX You may merge the main branch again because of 9135a1f41192ebe1d835282a1465dc284359d95c

**Pull request merged**